zeebe/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java

```diff
@@ -546,12 +549,25 @@ public long setCommitIndex(final long commitIndex) {
     checkArgument(commitIndex >= 0, "commitIndex must be positive");
     final long previousCommitIndex = this.commitIndex;
     if (commitIndex > previousCommitIndex) {
-      this.commitIndex = commitIndex;
       raftLog.setCommitIndex(Math.min(commitIndex, raftLog.getLastIndex()));
       if (isLeader()) {
         // leader counts itself in quorum, so in order to commit the leader must persist
-        raftLog.flush();
+        try {
+          raftLog.flush();
+        } catch (final FlushException e) {
+          if (log.isWarnEnabled()) {
+            log.warn(
+                "Failed to flush commit at index %s, resetting journal to %s and stepping down"
+                    .formatted(commitIndex, previousCommitIndex),
+                e);
+          }
+          raftLog.setCommitIndex(previousCommitIndex);
+          transition(Role.FOLLOWER);
```

Comment:
Oh, so we truncate on becoming follower, right? So it should be fine. As long as we just keep trying to become follower if this doesn't work :smile:

Reply:
Now that I think about it...truncate could fail (I/O error). So we should make sure we retry instead of just going inactive. But that can be a follow up PR/issue in order to keep this small, and this is still a big improvement on the current behavior :+1: